state/caasmodel_test.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state_test
+
+import (
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/feature"
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/storage"
+	"github.com/juju/juju/testing/factory"
+)
+
+type CAASModelSuite struct {
+	ConnSuite
+}
+
+var _ = gc.Suite(&CAASModelSuite{})
+
+// createTestModelConfig returns a new model config and its UUID for testing.
+func (s *CAASModelSuite) createTestModelConfig(c *gc.C) (*config.Config, string) {
+	return createTestModelConfig(c, s.modelTag.Id())
+}
+
+func (s *CAASModelSuite) newCAASModel(c *gc.C) (*state.Model, names.ModelTag, names.UserTag) {
+	cfg, uuid := s.createTestModelConfig(c)
+	modelTag := names.NewModelTag(uuid)
+	owner := names.NewUserTag("test@remote")
+	model, st, err := s.State.NewModel(state.ModelArgs{
+		Type:      state.ModelTypeCAAS,
+		CloudName: "dummy",
+		Config:    cfg,
+		Owner:     owner,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	s.AddCleanup(func(*gc.C) { st.Close() })
+	return model, modelTag, owner
+}
+
+func (s *CAASModelSuite) TestNewModel(c *gc.C) {
+	s.SetFeatureFlags(feature.CAAS)
+
+	model, modelTag, owner := s.newCAASModel(c)
+	c.Assert(model.Type(), gc.Equals, state.ModelTypeCAAS)
+	c.Assert(model.UUID(), gc.Equals, modelTag.Id())
+	c.Assert(model.Tag(), gc.Equals, modelTag)
+	c.Assert(model.ControllerTag(), gc.Equals, s.State.ControllerTag())
+	c.Assert(model.Owner(), gc.Equals, owner)
+	c.Assert(model.Name(), gc.Equals, "testing")
+	c.Assert(model.Life(), gc.Equals, state.Alive)
+	c.Assert(model.CloudRegion(), gc.Equals, "")
+}
+
+func (s *CAASModelSuite) TestModelDestroy(c *gc.C) {
+	s.SetFeatureFlags(feature.CAAS)
+
+	model, _, _ := s.newCAASModel(c)
+	err := model.Destroy(state.DestroyModelParams{})
+	c.Assert(err, jc.ErrorIsNil)
+	err = model.Refresh()
+	c.Assert(err, jc.ErrorIsNil)
+	// TODO(caas) - this will be dying when we add cleanup steps.
+	c.Assert(model.Life(), gc.Equals, state.Dead)
+}
+
+func (s *CAASModelSuite) TestCAASModelsCantHaveCloudRegion(c *gc.C) {
+	s.SetFeatureFlags(feature.CAAS)
+	cfg, _ := s.createTestModelConfig(c)
+	_, _, err := s.State.NewModel(state.ModelArgs{
```

Comment:
Can we please change NewModel to two new methods, NewIAASModel and NewCAASModel, and leave CloudRegion out of the latter? Rename the existing NewModel to NewIAASModel and drop Type from the ModelArgs (also renamed to IAASModelArgs). Then add a new method/type.